enterprise/server/remote_execution/runner/runner.go

```diff
@@ -374,7 +375,37 @@ func (r *commandRunner) Run(ctx context.Context) *interfaces.CommandResult {
 		return r.sendPersistentWorkRequest(ctx, command)
 	}
 
-	return r.Container.Exec(ctx, command, &commandutil.Stdio{})
+	execResult := r.Container.Exec(ctx, command, &commandutil.Stdio{})
+
+	// If the runner has exceeded 90% of memory or disk usage, don't try to recycle
+	// it, because that may cause failures if it's reused
+	maxedOutStr := ""
+	for _, fsUsage := range execResult.UsageStats.GetPeakFileSystemUsage() {
+		if float64(fsUsage.UsedBytes)/float64(fsUsage.TotalBytes) >= .9 {
+			maxedOutStr += fmt.Sprintf(" %d/%d B disk used for %s", fsUsage.UsedBytes, fsUsage.TotalBytes, fsUsage.GetSource())
+		}
+	}
+	// Only firecracker containers set max memory limits on the VMs
+	if fc, ok := r.Container.Delegate.(*firecracker.FirecrackerContainer); ok {
+		usedMemoryBytes := execResult.UsageStats.MemoryBytes
```

Comment:
while testing this locally I got a nil pointer exception at this line, so I think UsageStats is not guaranteed to be set. That's sort of an issue in and of itself which I'm digging into, but for now we should use `GetMemoryBytes()` to avoid the nil dereference